### PR TITLE
give OutputTransformer a proper lexer

### PR DESCRIPTION
**Pull request by pongad**

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 1]
```diff
@@ -359,4 +357,59 @@ private ScopeTable newChild() {
...
+      p += Character.charCount(cp);
+      return cp;
+    }
+
+    private static boolean identLead(int codePoint) {
+      return Character.isLetter(codePoint) || "$_".indexOf(codePoint) >= 0;
+    }
+
+    private static boolean identFollow(int codePoint) {
+      return identLead(codePoint) || Character.isDigit(codePoint);
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: "$" handling: We want `$` to be an identifier modifier prefix for our own purposes (eg `$resp`). Specifically, my understanding is that we were going to allow it _just_ to be a prefix. If that's not what you had in mind, let's discuss. 
If it is the case, then `$` should not be allowed in this function. As you have it, this allows identifiers such as `$$`, `$foo$bar`, and `$foo$`.

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 2]
```diff
@@ -359,4 +357,59 @@ private ScopeTable newChild() {
       return new ScopeTable(this);
     }
   }
+
+  private static class Scanner {
+    private static final int EOF = -1;
+    private static final int IDENT = -2;
+
+    private final String s;
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: `s` -> `str` or `input`

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 3]
```diff
@@ -359,4 +357,59 @@ private ScopeTable newChild() {
...
+    private String token = "";
+    private int p;
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: s/`p`/`nextLoc`/

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 4]
```diff
@@ -359,4 +357,59 @@ private ScopeTable newChild() {
...
+
+    private Scanner(String s) {
+      this.s = s;
+    }
+
+    private int scan() {
+      token = "";
+      int cp;
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: s/`cp`/`currCh`/ or similar

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 5]
```diff
@@ -359,4 +357,59 @@ private ScopeTable newChild() {
       return new ScopeTable(this);
     }
   }
+
+  private static class Scanner {
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: This function should have a test.

Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: A brief function comment would be useful, saying what characters we accept, and that `scan()` returns `EOF`, `IDENT`, or a character, and that in the case of `IDENT`, the token can be fetched via `.token`

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 6]
```diff
@@ -181,7 +186,7 @@ public void run() {
    * </code></pre>
    */
   private static OutputView.VariableView accessorNewVariable(
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: Do we have a test for this function?

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java (hunk 4 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: Changed to codePoint, cp is not a char.

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java (hunk 6 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: I think most of these are just tested by baselines.

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java (hunk 5 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: Done

Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java (hunk 1 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: Done. Note that `$foo$bar` should then be legal. The string consists of two identifiers, `$foo` and `$bar`.

File: src/test/java/com/google/api/codegen/transformer/ScannerTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    Scanner scanner = new Scanner("$a$b$");
+
+    assertThat(scanner.scan()).isEqualTo(Scanner.IDENT);
+    assertThat(scanner.token()).isEqualTo("$a");
+
+    assertThat(scanner.scan()).isEqualTo(Scanner.IDENT);
+    assertThat(scanner.token()).isEqualTo("$b");
+
+    assertThrow(() -> scanner.scan());
+  }
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/transformer/ScannerTest.java]: I would also test that `"a$$b"`  and `"a$"` throw

File: src/main/java/com/google/api/codegen/transformer/OutputTransformer.java [hunk 8]
```diff
@@ -360,54 +360,81 @@ private ScopeTable newChild() {
...
-    private static final int EOF = -1;
-    private static final int IDENT = -2;
+  /**
+   * Tokenizes a string. {@code scan()} returns the next token. A token is either the next unicode
+   * code point in the string or a special token. If a special token is returned, the string value
+   * of the token can be retrieved from {@code token}.
+   *
+   * <p>Special tokens:
+   *
+   * <pre>{@code
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/transformer/OutputTransformer.java]: It does, but `<pre>` preserves them :)

**Pull request merged**